torchtune/models/phi3/_sentencepiece.py

```diff
@@ -0,0 +1,224 @@
+                Message(role="assistant", content="assistant response\n"),
+                ]
+            # tokenize_messages encodes messages separately and concats
+            >>> tokenizer.tokenize_messages(messages, max_seq_len)[0]
+            [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+            # Same result as encoding the full string in one go
+            >>> tokenizer.encode(''.join([message.content for message in messages]))
+            [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+        Args:
+            messages (List[Message]): A list of messages, each containing role, content,
+                and masked attributes.
+            max_seq_len (Optional[int]): A max sequence length to truncate tokens to.
+                Default: None
+            add_eos (bool): Whether to append EOS after assistant message, default to False
+            ignore_system_prompts (bool): Whether to ignore system prompts, default to True.
+
+        Raises:
+            ValueError: If the role is not "user", "assistant", or "system".
+
+        Returns:
+            Tuple[List[int], List[bool]]: The tokenized messages
+        """
+        start_of_turn = True
+        end_of_turn = False
+        prev_ends_with_space = False
+        tokenized_messages = []
+        mask = []
+
+        new_line_token_id = self.encode("\n", add_bos=False, add_eos=False)
+
+        for message in messages:
+            # Prepend BOS on start of new turns
+            if start_of_turn:
+                tokenized_messages.append(self.bos_id)
+                mask.append(message.masked)
+
+            # Add special tokens
+            if message.role == "user":
+                tokenized_messages.append(self.special_tokens["<|user|>"])
+                mask.append(message.masked)
+            elif message.role == "assistant":
+                tokenized_messages.append(self.special_tokens["<|assistant|>"])
+                # If assistant message, this is the end of a turn
+                end_of_turn = message.role == "assistant"
+                mask.append(message.masked)
+            elif message.role == "system":
+                if ignore_system_prompts:
+                    continue
+                else:
+                    tokenized_messages.append(self.special_tokens["<|system|>"])
+                    mask.append(message.masked)
+            else:
+                raise ValueError(
+                    f"Unknown role '{message.role}' for message: '{message.content}'"
+                )
+
+            # Add new line token
+            tokenized_messages.extend(new_line_token_id)
+            mask.append(message.masked)
+
+            # Tokenize current message, append with masks
+            tokens = self.encode(
+                message.content.rstrip(" "),
+                add_bos=False,
+                add_eos=False,
+                trim_leading_whitespace=True,
```

Comment:
Always trimming whitespace (just to match HF tokenizer implementation)

Reply:
And this